### PR TITLE
Feat/node upgrade hook

### PR DESCRIPTION
**Pull request by philipakash**
https://github.com/openshift/openshift-ansible/pull/7736

### CODE REVIEW

File: playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
 - name: Drain and upgrade nodes
   hosts: oo_nodes_to_upgrade:!oo_masters_to_config
   # This var must be set with -e on invocation, as it is not a per-host inventory var
   # and is evaluated early. Values such as "20%" can also be used.
   serial: "{{ openshift_upgrade_nodes_serial | default(1) }}"
   max_fail_percentage: "{{ openshift_upgrade_nodes_max_fail_percentage | default(0) }}"
-
+  roles:
+  - lib_openshift
+  - openshift_facts
```
Comment (vrutkovs) [inline playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml]: This role seems to be called twice, is it required?

File: playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml [hunk 2]
```diff
@@ -23,35 +29,59 @@
     retries: 10
     delay: 5
     register: node_unschedulable
-    until: node_unschedulable|succeeded
+    until: node_unschedulable is succeeded
 
   - name: Drain Node for Kubelet upgrade
     command: >
-      {{ hostvars[groups.oo_first_master.0].openshift.common.admin_binary }} drain {{ openshift.node.nodename | lower }} --config={{ openshift.common.config_base }}/master/admin.kubeconfig --force --delete-local-data --ignore-daemonsets
+      {{ hostvars[groups.oo_first_master.0]['first_master_client_binary'] }} adm drain {{ openshift.node.nodename | lower }}
```
Comment (vrutkovs) [inline playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml]: This change doesn't seem to be related to node hooks, is it required?

Comment (philipakash) [reply playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml]: Feat/nodeupgradehooks 3.9 was good ... will reverify the code and fix it :-)

**Comment by philipakash** [comment]
@vrutkovs updated the correct changes

File: playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml [hunk 3]
```diff
@@ -12,6 +12,13 @@
       name: lib_openshift
     static: yes
 
+  # Run the pre-upgrade hook if defined:
+  - debug: msg="Running node pre-upgrade hook {{ openshift_node_upgrade_pre_hook }}"
+    when: openshift_node_upgrade_pre_hook is defined
+
+  - include_tasks: "{{ openshift_node_upgrade_pre_hook }}"
```
Comment (vrutkovs) [inline playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml]: `include_tasks` is not allowed to be used in Ansible 2.3 (which is required for 3.7). `include` should be used instead

**Comment by vrutkovs** [comment]
bot, test pull request

File: playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml (hunk 3 above)
Comment (philipakash) [reply playbooks/common/openshift-cluster/upgrades/upgrade_nodes.yml]: hmm make sense

**Comment by philipakash** [comment]
@vrutkovs Do you see any issues ... I see a couple of failures but lint tests are all ok.

**Comment by vrutkovs** [comment]
The PR is fine, but CI for 3.7 is broken.

I'll merge this manually soon unless the CI gets better soon

**Comment by philipakash** [comment]
@vrutkovs Just a general question on the release ... What is the release cycle for minor changes on the      3.9 and 3.7 once the code changes are merged to master. Or I would love to learn if there is any doc  which describes the release cycle ?

**Comment by sdodson** [comment]
Backports #6816 (and other fixes)
